### DISCUSSION TITLE
Set a environment variable during pre-render step

### DISCUSSION THREAD

**Discussion by boshek** [quarto-dev/quarto-cli#9195] [Q&A] [answered]
### Description

I'd like to figure out how to set an environment variable with quarto _during the pre-render_ step. I've tried two approaches so far [here](https://github.com/boshek/quarto-env-var-reprex). 

### Create a new file
I thought maybe I could [create pre-render](https://github.com/boshek/quarto-env-var-reprex/blob/8b1203ab3f64955e006beb689bca0f036d55791a/pre-process.R#L1-L4) a new file called `_environment-dynamic`. I wasn't sure if quarto would automatically pick up a bespoke `_environment-*` file but apparently not.

### Append a variable to `_environment` 
I also tried[ temporarily adding a line](https://github.com/boshek/quarto-env-var-reprex/blob/8b1203ab3f64955e006beb689bca0f036d55791a/pre-process.R#L6-L8) to the `_environment` file and then [remove it](https://github.com/boshek/quarto-env-var-reprex/blob/8b1203ab3f64955e006beb689bca0f036d55791a/post-process.R#L3-L8) in post render

Neither of those seemed to work. My approach to now was to simply create a quarto doc that defined the var and then added that to each qmd in the website like [here](https://github.com/boshek/quarto-env-var-reprex/blob/8b1203ab3f64955e006beb689bca0f036d55791a/set-vars.qmd#L5)

### Question
I am trying to move up to call stack and set that variable (dynamically) earlier in the pre-process.R file. Does anyone have any suggestions on how I might accomplish setting an environment variable that can be used by any of the pages in a website? 

**quarto version**: `1.4.549`
**os: macOS**:  `14.4.1`

**Comment by cderv** [comment, marked as answer]
I don't think this is a use case we have think of from pre-render step to modify the environment. 

Since 1.5 we've made changes so that project context is reloaded after pre-render step. Loading `_environment` is part of the project context. 

So if you try 1.5 (pre-release) it should work. (https://quarto.org/docs/download/prerelease.html)

Though I notice `quarto preview` on this project may not like that, because preview server stops when environment changes 

````
environment variables changed: preview terminating
````

we may need to tweak this 🤔

**Reply by boshek** [reply]
Ah yes `1.5` _does_ work! But as you say, it also does still kill the preview server which ends up being a bigger problem that the one we are discussing. Is it helpful to open an issue for this?

**Reply by cderv** [reply]
Already done at 
* https://github.com/quarto-dev/quarto-cli/issues/9199

**Reply by boshek** [reply]
Marked your upgrade suggestion as the answer then. Thanks!
